service/src/main/java/bio/terra/pipelines/service/PipelinesService.java

```diff
@@ -382,4 +386,73 @@ public List<WorkflowOutputDefinition> prepareCbasWorkflowOutputRecordUpdateDefin
             })
         .toList();
   }
+
+  /**
+   * Retrieve and format the pipeline inputs for a pipeline. Combine the user-provided inputs with
+   * the service-provided inputs to create a map of all the inputs for the pipeline. Apply the
+   * following customizations: - use custom (environment-specific) values for certain
+   * service-provided inputs - prepend the storage workspace container URL to the service-provided
+   * inputs that need it - prepend the control workspace container URL to the user-provided file
+   * inputs
+   *
+   * @param jobId UUID
+   * @param allInputDefinitions List<PipelineInputDefinition>
+   * @param userProvidedPipelineInputs Map<String, Object>
+   * @param controlWorkspaceContainerUrl String
+   * @param inputsWithCustomValues Map<String, Object> from pipeline Configuration
+   * @param keysToPrependWithStorageWorkspaceContainerUrl List<String> from pipeline Configuration
+   * @param storageWorkspaceContainerUrl String from pipeline Configuration
+   * @return formattedPipelineInputs Map<String, Object>
+   */
+  public Map<String, Object> formatPipelineInputs(
+      UUID jobId,
+      List<PipelineInputDefinition> allInputDefinitions,
+      Map<String, Object> userProvidedPipelineInputs,
+      String controlWorkspaceContainerUrl,
+      Map<String, Object> inputsWithCustomValues,
+      List<String> keysToPrependWithStorageWorkspaceContainerUrl,
+      String storageWorkspaceContainerUrl) {
+
+    Map<String, Object> allPipelineInputs =
+        constructRawInputs(allInputDefinitions, userProvidedPipelineInputs);
+    List<String> userProvidedInputFileKeys = extractUserProvidedFileInputNames(allInputDefinitions);
+    Map<String, Object> formattedPipelineInputs = new HashMap<>();
+
+    for (PipelineInputDefinition inputDefinition : allInputDefinitions) {
+      String keyName = inputDefinition.getName();
+      String wdlVariableName = inputDefinition.getWdlVariableName();
+      PipelineVariableTypesEnum pipelineInputType = inputDefinition.getType();
+
+      // if value not present in allPipelineInputs and the input is optional, continue
+      if (!inputsWithCustomValues.containsKey(keyName)
+          && !allPipelineInputs.containsKey(keyName)
+          && !inputDefinition.isRequired()) {
+        continue;
```

Comment:
currently we don't add default values for optional user-provided inputs to the pipeline inputs. (we don't have optional inputs in the imputation pipeline.) should we add those if they're present?

Reply:
yea i think that makes sense, every optional user-provided input should have a default in our database and that should be used here.